src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -14,115 +13,150 @@ export interface EventDetails {
     eventArgs: any[];
 }
 
+export interface DeferredEventDetails extends EventDetails {
+    deferredResolution: ExternalResolutionPromise;
+}
 export interface AdapterListener {
-    (eventType: string): (eventArgs: any[]) => any;
+    (...eventArgs: any[]): any;
 }
 
 const TWO_MINUTES = 120000;
 const FOUR_MINUTES = 2 * TWO_MINUTES;
 
+// BrowserAdapterEventManager is to be used by a BrowserAdapter to ensure the browser does not
+// determine that the service worker can be shut down due to events not responding within 5 minutes.
+//
+// It is responsible for acting as a mediator between browser-level events and application-level listeners:
+//   * registering a handleEvent listener per eventType to respond to all browser-level events by:
+//      * ensuring a response is sent to the browser within 4 minutes to prevent the 5-minute watchdog timeout
+//      * delaying potential fire-and-forget responses for 2 minutes to give potential promises time to finish
+//      * deferring any events without an associated application-level listener until one is registered
+//   * registering application-level listeners to be called inside the handleEvent listener
+
 export class BrowserAdapterEventManager {
-    protected deferredEvents: EventDetails[] = [];
-    protected eventsToApplicationListenersMapping: DictionaryStringTo<ApplicationListener[]> = {};
-    protected adapterListener: AdapterListener =
+    protected deferredEvents: DeferredEventDetails[] = [];
+    protected eventsToApplicationListenersMapping: DictionaryStringTo<ApplicationListener> = {};
+    protected eventsToAdapterListenersMapping: DictionaryStringTo<AdapterListener> = {};
+    protected handleEvent: (eventType: string) => AdapterListener =
         (eventType: string) =>
         (...eventArgs: any[]) => {
-            this.processEvent(eventType, eventArgs);
+            const processedResults = this.processEvent(eventType, eventArgs); //guard in case we return undefined for a timed out promise resoponse
```

Comment:
I'm not sure I understand the comment here; the new behavior for the 4 minute timeout promise turns a "timeout" from a event that rejects the promise to an event that resolves the Promise as undefined, but the value that `processedResults` gets set to here is still going to be a Promise, not `undefined`.